test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs

```diff
@@ -76,6 +76,70 @@ public void Save_NullSession_NonEmptyDictionary_Throws()
             });
         }
 
+        [Theory]
+        [MemberData(nameof(InvalidTypes))]
+        public void EnsureObjectCanBeSerialized_InvalidType_Throws(string key, object value, Type type)
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+
+            // Act & Assert
+            var exception = Assert.Throws<InvalidOperationException>(() =>
+            {
+                testProvider.EnsureObjectCanBeSerialized(new Dictionary<string, object> { { key, value } });
+            });
+            Assert.Equal($"The type {type} cannot be serialized to Session by '{typeof(SessionStateTempDataProvider).FullName}'.",
+                exception.Message);
+        }
+
+        [Theory]
+        [MemberData(nameof(ValidTypes))]
+        public void EnsureObjectCanBeSerialized_ValidType_DoesNotThrow(string key, object value)
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+
+            // Act & Assert (Does not throw)
+            testProvider.EnsureObjectCanBeSerialized(new Dictionary<string, object> { { key, value } });
+        }
+
+        public static TheoryData<string, object, Type> InvalidTypes
+        {
+            get
+            {
+                return new TheoryData<string, object, Type>
+                {
+                    { "Object", new object(), typeof(object) },
+                    { "ObjectArray", new object[3], typeof(object) },
+                    { "TestItem", new TestItem(), typeof(TestItem) },
+                    { "ListTestItem", new List<TestItem>(), typeof(TestItem) },
+                    { "DictTestItem", new Dictionary<string, TestItem>(), typeof(TestItem) }
```

Comment:
Maybe also add `Dictionary<object, string>` and `Dictionary<TestItem, TestItem>` here